### PR TITLE
Response mode restriction for FAPI code flow

### PR DESCRIPTION
**Pull request by RivinduM**
### Proposed changes in this pull request

According to the FAPI compliance, the authorization server must require the response mode to be jwt if the response type is code[1]. With this PR, the response mode jwt is validated as a mandatory parameter for the code flow if the application is FAPI compliant.

[1] https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server

### Related Issues
https://github.com/wso2/product-is/issues/16513

### CODE REVIEW

**Review by Akila94 — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2004,6 +2004,14 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
             validateNonceParameter(params.getNonce());
         }
 
+        try {
+            if (OAuth2Util.isFapiConformantApp(params.getClientId())) {
+                EndpointUtil.validateFAPIResponseMode(params.getResponseType(), params.getResponseMode());
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthSystemException(e);
```
Comment (lalaji) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2012]: no error message added to throw clause?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 2]
```diff
@@ -232,7 +234,11 @@ private void validateAuthzRequest(HttpServletRequest request) throws ParCoreExce
 
         try {
             OAuthAuthzRequest oAuthAuthzRequest = getOAuthAuthzRequest(request);
-            validateRequestObject(oAuthAuthzRequest);
+            RequestObject requestObject = validateRequestObject(oAuthAuthzRequest);
+            if (isFAPIConformantApp(oAuthAuthzRequest.getClientId())) {
+                EndpointUtil.validateFAPIResponseMode(oAuthAuthzRequest.getResponseType(),
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:239]: write a similar method to get the response type as well overriding the req obj parameters

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 3]
```diff
@@ -1899,4 +1901,20 @@ private static Class<? extends OAuthAuthzRequest> getOAuthAuthzRequestClass() th
...
+    /**
+     * Validate the response mode against the response type as per FAPI spec.
+     *
+     * @param responseType response mode
+     * @param responseMode response type
+     * @throws OAuthProblemException when response mode is not valid
+     */
+    public static void validateFAPIResponseMode(String responseType, String responseMode)
+            throws OAuthProblemException {
+
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1914]: Add comment

Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1916]: Fixed with f22e2c736c5a4089d66f1f1e2c2c2c5cc6c1930a

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java (hunk 2 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java]: fixed with f22e2c736c5a4089d66f1f1e2c2c2c5cc6c1930a

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -4183,4 +4187,12 @@ private void addUserAttributesToCache(SessionDataCacheEntry sessionDataCacheEntr
                 new DeviceAuthorizationGrantCacheEntry(sessionDataCacheEntry.getLoggedInUser().getUserAttributes());
         DeviceAuthorizationGrantCache.getInstance().addToCache(cacheKey, cacheEntry);
     }
+
+    private boolean isFapiConformant(String clientId) throws InvalidRequestException {
+        try {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4192]: Add new line after method definition

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java [hunk 5]
```diff
@@ -235,14 +238,56 @@ private void validateAuthzRequest(HttpServletRequest request) throws ParCoreExce
 
         try {
             OAuthAuthzRequest oAuthAuthzRequest = getOAuthAuthzRequest(request);
-            validateRequestObject(oAuthAuthzRequest);
+            RequestObject requestObject = validateRequestObject(oAuthAuthzRequest);
+            Map<String, String> oauthParams = overrideRequestObjectParams(request, requestObject);
+            if (isFAPIConformantApp(oAuthAuthzRequest.getClientId())) {
+                EndpointUtil.validateFAPIResponseMode(oauthParams.get(RESPONSE_TYPE), oauthParams.get(RESPONSE_MODE));
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/par/OAuth2ParEndpoint.java:244]: lets change the method name in the util to give more clarity. Current method name reads like that there is a response mode named "FAPI"
```suggestion
                EndpointUtil.validateFAPIAllowedResponseMode(oauthParams.get(RESPONSE_TYPE), oauthParams.get(RESPONSE_MODE));
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 6]
```diff
@@ -1899,4 +1901,22 @@ private static Class<? extends OAuthAuthzRequest> getOAuthAuthzRequestClass() th
...
+    /**
+     * Validate the response mode against the response type as per FAPI spec.
+     * Response mode required to be jwt if code response type is used.
+     * <a href="https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server">5.2.2-2.2</a>
+     *
+     * @param responseType response mode
+     * @param responseMode response type
+     * @throws OAuthProblemException when response mode is not valid
+     */
+    public static void validateFAPIResponseMode(String responseType, String responseMode)
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1914]: lets change the method name in the util to give more clarity. Current method name reads like that there is a response mode named "FAPI"

```suggestion
    public static void validateFAPIAllowedResponseMode(String responseType, String responseMode)
```

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1914]: Also doesn't the spec say to validate having `code id_token` in https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server `2.1`?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 7]
```diff
@@ -2014,6 +2014,10 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
             validateNonceParameter(params.getNonce());
         }
 
+        if (isFapiConformant(params.getClientId())) {
+            EndpointUtil.validateFAPIResponseMode(params.getResponseType(), params.getResponseMode());
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2018]: Also doesn't the spec say to validate having `code id_token` in https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server `2.1`?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 6 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1914]: Also there should be a unit test for this.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4193]: Fixed in 9ed19436be3c36b2fe703d66e088386ce1c78c86

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 6 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Fixed in 9ed19436be3c36b2fe703d66e088386ce1c78c86

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 7 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed in a9450470a49c23ea24851d9fb636453829c624e9

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 8]
```diff
@@ -1899,4 +1902,22 @@ private static Class<? extends OAuthAuthzRequest> getOAuthAuthzRequestClass() th
...
+    /**
+     * Validate the response mode against the response type as per FAPI spec.
+     * Response mode required to be jwt if code response type is used.
+     * <a href="https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server">5.2.2-2.2</a>
+     *
+     * @param responseType response mode
+     * @param responseMode response type
+     * @throws OAuthProblemException when response mode is not valid
+     */
+    public static void validateFAPIAllowedResponseMode(String responseType, String responseMode)
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1915]: Lets rename the method to better fit the validations done. Maybe something like `validateFAPIAllowedResponseTypeAndMode`
```
shall require; 
1. the response_type value code id_token, or
2. the response_type value code in conjunction with the response_mode value jwt;
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 9]
```diff
@@ -1899,4 +1902,22 @@ private static Class<? extends OAuthAuthzRequest> getOAuthAuthzRequestClass() th
...
+            throws OAuthProblemException {
+
+        if (!(CODE_IDTOKEN.equals(responseType) || (CODE.equals(responseType) && JWT.equals(responseMode)))) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1918]: isnt other types of JWT response modes accepted? (ex: query.jwt, fragment.jwt, form_post.jwt)

In the JARM spec it states `The default for response type "code" is "query.jwt"`. That means using `query.jwt` response mode in conjunction with `code` response type is similar to using code + jwt combination.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 8 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1915]: Additionally, there should be a unit test to make sure the spec allowed values are covered.

Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: fixed with [a2e92b8](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2157/commits/a2e92b846147834a7532ac6bfe3df2bb9bc6c8b2)

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 9 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1920]: In all the places the FAPI spec explicitly mentions that it requires response mode value to be jwt when code response type is used.

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1920]: Can we check this against the conformance suite and also reach out to the FAPI working group for guidance? I agree the spec mentions that, but it doesn't seem logical.

Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:1920]: The following test is in the conformance suite
```
FAPI1-Advanced-Final: ensure response_type code fails

"This test uses response_type=code in the authorization request, which (as a JARM response has not been requested)
 is not permitted in FAPI1-Advanced - only the hybrid flow ('response_type=code id_token') or JARM 
 ('response_type=code&response_mode=jwt') is allowed. The authorization server should show an error message that
  the response type is unsupported or the request is invalid (a screenshot of which should be uploaded) or the user 
  should be redirected back to the conformance suite with a correct error response, or an error could be returned from the 
  PAR endpoint.",
  ```

**Review by janakamarasena — DISMISSED** [review]

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**